tempesta_fw/http_msg.c

```diff
@@ -26,6 +26,67 @@
 #include "http_msg.h"
 #include "ss_skb.h"
 
+/**
+ * Build TfwStr repesented HTTP header.
+ * @name	- header name without ':';
+ * @value	- header value;
+ */
+TfwStr *
+tfw_http_msg_make_hdr(TfwPool *pool, const char *name, const char *val)
+{
+	size_t n_len = strlen(name);
+	size_t v_len = val ? strlen(val) : 0;
+	const TfwStr tmp_hdr = {
+		.ptr = (TfwStr []){
+			{ .ptr = (void *)name,	.len = n_len },
+			{ .ptr = S_DLM,		.len = SLEN(S_DLM) },
+			{ .ptr = (void *)val,	.len = v_len },
+		},
+		.len = n_len + SLEN(S_DLM) + v_len,
+		.eolen = 2,
+		.flags = (val ? 3 : 2) << TFW_STR_CN_SHIFT
+	};
+
+	return tfw_strdup(pool, &tmp_hdr);
+}
+EXPORT_SYMBOL(tfw_http_msg_make_hdr);
+
+/**
+ * Get header id in header table for header @hdr.
+ */
+unsigned int
+tfw_http_msg_is_spec_hdr(const TfwStr *hdr)
+{
+	static const TfwStr spec_hdrs[] = {
+	#define TfwStr_string(v) { (v), NULL, sizeof(v) - 1, 0 }
+		[TFW_HTTP_HDR_HOST]	= TfwStr_string("Host:"),
+		[TFW_HTTP_HDR_CONTENT_LENGTH] = TfwStr_string("Content-Length:"),
+		[TFW_HTTP_HDR_CONTENT_TYPE] = TfwStr_string("Content-Type:"),
+		[TFW_HTTP_HDR_SERVER]	= TfwStr_string("Server:"),
+		[TFW_HTTP_HDR_COOKIE]	= TfwStr_string("Cookie:"),
+		[TFW_HTTP_HDR_ETAG]	= TfwStr_string("ETag:"),
+		[TFW_HTTP_HDR_CONNECTION] = TfwStr_string("Connection:"),
+		[TFW_HTTP_HDR_X_FORWARDED_FOR] = TfwStr_string("X-Forwarded-For:"),
+		[TFW_HTTP_HDR_KEEP_ALIVE] = TfwStr_string("Keep-Alive:"),
+		[TFW_HTTP_HDR_TRANSFER_ENCODING] = TfwStr_string("Transfer-Encoding:"),
+	#undef TfwStr_string
+	};
+	DEFINE_TFW_STR(user_agent, "User-Agent:");
+	DEFINE_TFW_STR(if_none_match, "If-None-Match:");
+
+	int i;
+
+	for (i = 0; i < TFW_HTTP_HDR_RAW; ++i)
+		if (!tfw_stricmpspn(hdr, &spec_hdrs[i], ':'))
+			return i;
+	if (!tfw_stricmpspn(hdr, &user_agent, ':'))
+		return TFW_HTTP_HDR_USER_AGENT;
+	if (!tfw_stricmpspn(hdr, &if_none_match, ':'))
+		return TFW_HTTP_HDR_IF_NONE_MATCH;
```

Comment:
Actually there is `__tfw_http_msg_find_hdr()` function for binary search. But it requires alphabetically sorted array. That's not true for the array since original `tfw_http_hdr_t` is not sorted. The function is called only at configuration processing, so i've decided to keep it as clean as possible

Reply:
Ok, then just point out this in a comment.

My concern is that we have several library-like files (str.[ch], ss_skb.[ch], http_msg.[ch]) which contain logic useful in various situations. The function also looks like very useful in future, so it's necessary either to limit it's usage or write in in very solid way.